src/main/rascal/refactor/Rename.rsc

```diff
@@ -29,200 +29,200 @@ module refactor::Rename
+        }
 
-    // STORE
-    map[str, value] store = ();
-    solver.readStore = value(str key) { return store[key]; };
-    solver.writeStore = void(str key, value val) {
-        store[key] = val;
-    };
+        return tvar(|unknown:///|);
+    }
+    set[Message] getMessages() = {toMessage(m, getType) | m <- messages};
 
-    // RUN
-    map[loc, Tree] treeCache = ();
-    map[loc, TModel] modelCache = ();
+    // Edits
+    set[loc] editsSeen = {};
+    list[DocumentEdit] docEdits = [];
 
-    Tree getTree(loc l) {
-        if (l notin treeCache) {
-            treeCache[l] = config.parseLoc(l);
-        } else if (config.debug) {
-            println("-- Using cached tree for <l>");
+    void checkEdit(replace(loc range, _)) {
+        if (range in editsSeen) {
+            registerMessage(error(range, "Multiple replace edits for this location."));
         }
-        return treeCache[l];
+        editsSeen += range;
     }
 
-    TModel getTModel(loc l) {
-        if (l notin modelCache) {
-            modelCache[l] = config.tmodelForLoc(l);
-        } else if (config.debug) {
-            println("-- Using cached TModel for <l>");
+    void checkEdit(DocumentEdit e) {
+        loc file = e has file ? e.file : e.from;
+        if (changed(f, tes) := e) {
+            // Check contents of DocumentEdit
+            for (te:replace(range, _) <- tes) {
+                // Check integrity
+                if (range.top != f) {
+                    registerMessage(error(range, "Invalid replace edit for this location. This location is not in <f>, for which it was registered."));
+                }
+
+                // Check text edits
+                checkEdit(te);
+            }
+        } else if (file in editsSeen) {
+            registerMessage(error(file, "Multiple <getName(e)> edits for this file."));
         }
-        return modelCache[l];
+
+        editsSeen += file;
     }
 
-    solver.run = RenameResult() {
-        while (treeTaskQueue != [] || modelTaskQueue != []) {
-            treeTaskQueueCopy = treeTaskQueue;
-            modelTaskQueueCopy = modelTaskQueue;
+    void registerDocumentEdit(DocumentEdit e) {
+        checkEdit(e);
+        docEdits += e;
+    };
 
-            // We will do all tasks in the queue
-            treeTaskQueue = [];
-            modelTaskQueue = [];
+    void registerTextEdit(TextEdit e) {
+        checkEdit(e);
 
-            for (loc f <- treeTaskQueueCopy.file + modelTaskQueueCopy.file) {
-                fileTreeTasks = treeTaskQueueCopy[f];
-                if (config.debug) println("<size(fileTreeTasks)> tasks for tree of <f>");
+        loc f = e.range.top;
+        if ([*pre, changed(f, prev)] := docEdits) {
+            // If possible, merge with latest document edit
+            // TODO Just assign to docEdits[-1], once this issue has been solved:
+            // https://github.com/usethesource/rascal/issues/2123
+            docEdits = [*pre, changed(f, prev + e)];
```

Comment:
better to say: `c:changed(f, _)` and then `c[changes=f.changes + e]` such that externally defined keywords of `c` are not lost due to this merge.